src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseSign1Message.cs

```diff
@@ -0,0 +1,283 @@
+            if (content == null)
+            {
+                throw new ArgumentNullException(nameof(content));
+            }
+
+            if (key == null)
+            {
+                throw new ArgumentNullException(nameof(key));
+            }
+
+            byte[] encodedProtectedHeader = CreateEncodedProtectedHeader(KeyType.RSA, hashAlgorithm);
+            byte[] toBeSigned = CreateToBeSigned(SigStructureCoxtextSign1, encodedProtectedHeader, content);
+            byte[] signature = SignWithRSA(key, toBeSigned, hashAlgorithm);
+            return SignCore(encodedProtectedHeader, GetEmptyCborMap(), signature, content, isDetached);
+        }
+
+        [UnsupportedOSPlatform("browser")]
+        public static byte[] Sign(byte[] content, CoseHeaderMap protectedHeaders, CoseHeaderMap unprotectedHeaders, AsymmetricAlgorithm key, bool isDetached = false)
+        {
+            if (content == null)
+            {
+                throw new ArgumentNullException(nameof(content));
+            }
+
+            if (key == null)
+            {
+                throw new ArgumentNullException(nameof(key));
+            }
+
+            if (protectedHeaders == null)
+            {
+                throw new ArgumentNullException(nameof(protectedHeaders));
+            }
+
+            if (unprotectedHeaders == null)
+            {
+                throw new ArgumentNullException(nameof(unprotectedHeaders));
+            }
+
+            KeyType keyType = key switch
+            {
+                ECDsa => KeyType.ECDsa,
+                RSA => KeyType.RSA,
+                _ => throw new CryptographicException(SR.Format(SR.Sign1SignUnsupportedKey, key.GetType()))
+            };
+
+            ThrowIfDuplicateLabels(protectedHeaders, unprotectedHeaders);
+
+            ReadOnlyMemory<byte> encodedAlg = GetCoseAlgorithmFromBuckets(protectedHeaders, unprotectedHeaders);
+
+            int? algorithmHeader = DecodeCoseAlgorithmHeader(encodedAlg);
+            Debug.Assert(algorithmHeader.HasValue, "Algorithm (alg) is a known header and should have been validated in Set[Encoded]Value()");
+
+            HashAlgorithmName hashAlgorithm = GetHashAlgorithmFromCoseAlgorithmAndKeyType(algorithmHeader.Value, keyType);
+
+            byte[] encodedProtetedHeaders = protectedHeaders.Encode(mustReturnEmptyBstrIfEmpty: true);
+            byte[] toBeSigned = CreateToBeSigned(SigStructureCoxtextSign1, encodedProtetedHeaders, content);
+
+            byte[] signature = keyType == KeyType.ECDsa ?
+                SignWithECDsa((ECDsa)key, toBeSigned, hashAlgorithm) :
+                SignWithRSA((RSA)key, toBeSigned, hashAlgorithm);
+
+            return SignCore(encodedProtetedHeaders, unprotectedHeaders.Encode(), signature, content, isDetached);
+        }
+
+        // Validate duplicate labels https://datatracker.ietf.org/doc/html/rfc8152#section-3.
+        internal static void ThrowIfDuplicateLabels(CoseHeaderMap @protected, CoseHeaderMap unprotected)
```

Comment:
I'd call them protectedHeaders and unprotectedHeaders to avoid the `@`, personally.